File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        carry = 0
        dummy = ListNode(0)
        tail = dummy

        while l1 or l2 or carry:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0
            total = num1+num2 + carry 
            num3 =  total % 10 
            carry = total // 10

            node = ListNode(num3)
            tail.next = node
            tail = tail.next

            if l1 : l1 = l1.next
            if l2 : l2 = l2.next
        return dummy.next
  

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) {
        if(lists.length ==0 ) return null;

        for(let i=0;i<lists.length;i++)
            lists[i] = this.mergeLists(lists[i],lists[i-1]);
            
        return lists[lists.length-1];

    }

    mergeLists(l1,l2){
        let dummy = new ListNode(0);
        let curr = dummy;

        while(l1 && l2){
            if(l1.val > l2.val){
                curr.next = l2;
                l2 = l2.next;
            }else{
                curr.next = l1;
                l1 = l1.next;
            }
            curr = curr.next;
        }

        if(l1){
            curr.next = l1;
            curr = curr.next;
        }

        if(l2){
            curr.next = l2;
            curr = curr.next;
        }
        return dummy.next;
    }
}
